### ISSUE TITLE
Please support semantic versioning and adopt API stability

### ISSUE THREAD

**Issue by btipling** [rackt/react-router#2807]
It doesn't seem like react-router supports semantic versioning. From [the 2.0.0 upgrade guide](https://github.com/rackt/react-router/blob/master/upgrade-guides/v2.0.0.md): 

> If you drop in v2.x into a v1.x application and it doesn't run, then there is a bug.

In terms of [semantic versioning](http://semver.org/) that should not have resulted in a new major version. Also considering that I and others had problems with #2695, breaking tests and deployment with no code changes on our part, this is not just a one time problem. Please adopt semantic versioning to help developers out. 

It's been a rough 2 months for react-router, but I also hope that the project can adopt some long lasting API stability. Nothing more than about a month between major version change is bad for everyone.

**Comment by taion** [comment]
Semantic versioning permits us to increment the major version whenever we want. There are also technically some under-the-hood breaking changes, but those are on maintainers of integrations, not on direct consumers.

Deprecation warnings are generally accepted under "feature" releases. This is the upstream React plan for versioning post-1.x per https://gist.github.com/zpao/6e12ee0f46ce87af2287#versioning.

If you rely on a lack of deprecation warnings in development environments for your deployment to work, then you should be pinning to minor releases, not major releases - of all libraries involved, including transcluded peer dependencies like `history`.

**Comment by ryanflorence** [comment]
If we didn't provide backwards compatibility then this would obviously be a 2.0 release.

Because we provide backwards compatibility there is a question that semver *doesn't answer*.

We are following semver.

**Comment by ryanflorence** [comment]
> Also considering that I and others had problems with #2695, breaking tests and deployment with no code changes on our part, this is not just a one time problem

Which is **exactly why this is a 2.0 release**.

**Comment by ryanflorence** [comment]
The only question is "Does changing an API, but providing backwards compatibility with a warning, constitute a breaking change?"

You are saying "No" and "Yes" in your own comment here.

Because we provide backwards compat with warnings, you are saying "no" here:

> In terms of semantic versioning that should not have resulted in a new major version.

But because your build fails on console warnings of deprecated API, you are saying "yes".

> breaking tests and deployment with no code changes on our part, this is not just a one time problem

Providing backwards compatibility on breaking API changes is not answered by semver. Even you, in this issue, don't know what is supposed to happen because you are asking for both :P

**Comment by btipling** [comment]
> Semantic versioning permits us to increment the major version whenever we want.

A permission you certainly seem to be taking advantage of with a new major change just a month out, with little regard for the thousands of developers who depend on this library. Perhaps you might want to make a write up how version 1 was a failure and why 2 had to come so soon and what you will do in the future to avoid this kind of problem.

> Which is exactly why this is a 2.0 release.

Well given that you broke tests and deployments for us and probably others without us having a code change, you must admit that you should probably have then had the foresight to update versions prior to making breaking changes.

Perhaps there's a some finessing of words that can lend you to suggest you are following semantic versioning literally, but you certainly aren't following it in the spirit of making developer's dependency on react-router easy.

We're pretty dependent on this library and we just don't have the bandwidth to transition away and what would we even transition to, so I hope that the future for react-router is more stable. If not we will probably have move away. Losing days of productivity repeatedly has been hard to justify with project managers.

**Comment by taion** [comment]
While it's understandable that you're frustrated, I'd recommend investing some time in better familiarizing your team with npm.

We tried very hard to provide a smooth upgrade path, but fundamentally the only fix that was required for the deprecation warnings was to pin the dependency version on `history` to a release prior to the one that introduced the deprecation warnings.

This is actually quite easy to do - you just go into `package.json` and change `"history": "^1.12.0"` or the equivalent to something like `"history": "1.13.1"`.

We understand that React Router is a landing point for less experienced developers who are just learning React, but if you feel like you are deeply invested in the npm-based ecosystem, my best recommendation to you is to get sufficient familiarity with npm such that you understand the details of managing the versions of your dependencies.

npm is pretty much the best package manager I've ever worked with across many, many different languages and platforms, and I trust that you'll find it enjoyable and educational to learn how to use npm.

**Comment by ryanflorence** [comment]
I get that 0.13 -> 1.0 was a really painful upgrade.

> Well given that you broke tests and deployments for us and probably others without us having a version change, how does this support your argument?

Please remember that failing builds on warnings is an app choice, we didn't break your build. However, we respect that choice and that is exactly why we are versioning the way we are now when we deprecate API.

> with a new major version change just a month out. Was version 1 was a failure? Why did version 2 had have come so soon?

November 9th to January 9th will be 2 months. There were some integration points that we got wrong. I wish I knew how to write software perfectly the first time--or the 100th time for that matter--but I can't.

Moving forward, we will not release breaking API any faster than 3 months so people have time to upgrade.

> I hope that the future for react-router is more stable.

First, semver allows us to do whatever we want pre 1.0, and that's what we did.

Now that we're in the land of "stable versions", I don't know how to be more stable than what we're doing:

1. Release breaking API no faster than every 3 months
2. Provide complete backwards compatibility with the previous API for incremental code updates
3. Exhaustive console warnings for every changed API.
4. Providing a handful of codemods to automatically update your code.

What else could you expect from us?

(20 free hours of consulting I guess!)

I had no idea how much more time all of that takes, but it adds at least 3x. The 1.0 to 2.0 code changes would have been like 30 lines w/o all the stuff to make long-term usage of our library stable. I hope you can appreciate that.

I get that upgrading can be frustrating, and I appreciate your patience as we went from 0.13 to 1.0. I think you'll find our current strategy more than satisfactory and completely in line with semver.

**Comment by btipling** [comment]
@taion we pinned our version after the fact. It shouldn't have ever happened in the first place. None of the other dependencies we have or had had in the past had had this requirement. react-router broke things, and you are passing on blame to us? Do you recommend that everyone pin versions? This is not the default npm i --save behavior. Is that what you do? I only see one pinned versions in your own [package.json](https://github.com/rackt/react-router/blob/master/package.json), so you yourself would be susceptible to surprise breakage in the kind you introduced.

**Comment by ryanflorence** [comment]
Yeah, we screwed up our versioning with our history dependency. That is *exactly* why this is a 2.0 release. New API, with backward compat and warnings we consider to be breaking. Had we followed that with history you would be having a happier month.

I could be wrong, but it looks like you just want to vent (especially since you're asking us to do contradictory things) so I'm locking this conversation. Take it to Twitter :P
